gdnative-core/src/core_types/quat.rs

```diff
@@ -9,22 +10,274 @@ pub struct Quat {
     }
 
+    /// Constructs a quaternion from the given [Basis]
+    #[inline]
+    pub fn from_basis(basis: &Basis) -> Self {
+        basis.to_quat()
+    }
+
+    /// Constructs a quaternion that will perform a rotation specified by Euler angles (in the YXZ
+    /// convention: when decomposing, first Z, then X, and Y last), given in the vector format as
+    /// (X angle, Y angle, Z angle).
+    #[inline]
+    pub fn from_euler(euler: Vector3) -> Self {
+        Self::gd(glam::Quat::from_rotation_ypr(euler.y, euler.x, euler.z))
+    }
+
+    /// Constructs a quaternion that will rotate around the given axis by the specified angle. The
+    /// axis must be a normalized vector.
+    #[inline]
+    pub fn from_axis_angle(axis: Vector3, angle: f32) -> Self {
+        debug_assert!(axis.is_normalized(), "Axis is not normalized");
+        Self::gd(glam::Quat::from_axis_angle(axis.glam().into(), angle))
+    }
+
+    /// Performs a cubic spherical interpolation between quaternions `pre_a`, this quaternion, `b`,
+    /// and `post_b`, by the given amount `t`.
+    #[inline]
+    pub fn cubic_slerp(self, b: Self, pre_a: Self, post_b: Self, t: f32) -> Self {
+        let t2 = (1.0 - t) * t * 2.0;
+        let sp = self.slerp(b, t);
+        let sq = pre_a.slerpni(post_b, t);
+        sp.slerpni(sq, t2)
+    }
+
+    /// Returns the dot product of two quaternions.
+    #[inline]
+    pub fn dot(self, b: Self) -> f32 {
+        self.glam().dot(b.glam())
+    }
+
+    /// Returns Euler angles (in the YXZ convention: when decomposing, first Z, then X, and Y last)
+    /// corresponding to the rotation represented by the unit quaternion. Returned vector contains
+    /// the rotation angles in the format (X angle, Y angle, Z angle).
+    #[inline]
+    pub fn to_euler(self) -> Vector3 {
+        Basis::from_quat(self).to_euler()
+    }
+
+    /// Returns the inverse of the quaternion.
+    #[inline]
+    pub fn inverse(self) -> Self {
+        Self::gd(self.glam().inverse())
+    }
+
+    /// Returns `true` if this quaternion and `quat` are approximately equal, by running
+    /// `is_equal_approx` on each component
     #[inline]
-    pub fn is_equal_approx(self, to: &Self) -> bool {
+    pub fn is_equal_approx(self, to: Self) -> bool {
         self.x.is_equal_approx(to.x)
             && self.y.is_equal_approx(to.y)
             && self.z.is_equal_approx(to.z)
             && self.w.is_equal_approx(to.w)
     }
 
+    /// Returns whether the quaternion is normalized or not.
+    #[inline]
+    pub fn is_normalized(self) -> bool {
+        self.glam().is_normalized()
+    }
+
+    /// Returns the length of the quaternion.
+    #[inline]
+    pub fn length(self) -> f32 {
+        self.glam().length()
+    }
+
+    /// Returns the length of the quaternion, squared.
+    #[inline]
+    pub fn length_squared(self) -> f32 {
+        self.glam().length_squared()
+    }
+
+    /// Returns a copy of the quaternion, normalized to unit length.
```

Comment:
Maybe add to documentation something like:

> A normalized representation is necessary before transforming vectors through `xform()` or `*`.

Reply:
I think the documentation on just `mul` and `xform` should be sufficient. I would expect people to go straight to those functions anyways if they encounter any issues with it.

Reply:
Yeah, I think it can't hurt to mention the use case here too, so like it is now is good 🙂